### PR TITLE
stats servlet + stats HTML

### PR DESCRIPTION
**Pull request by GigaMatt**
Statistics steps 1-4 completed. I'm having issues retrieving the 2 messages the webpage says I should be noticing. I think this is due to my not exactly knowing how to login to the webpage through the local host.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (dgalbraith33) [inline .vscode/settings.json]: We shouldn't commit this file to the repo. I can walk you through how to delete it in chat.

You can add a line that says ".vscode/" in the ".gitignore" file so git won't recommend you add it in the future.

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 2]
```diff
@@ -27,57 +28,69 @@
 import java.util.List;
 import java.util.UUID;
 
+
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/data/Datastore.java]: Nit: delete this extra line.

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 3]
```diff
@@ -27,57 +28,69 @@
...
+  /** 
+   * Retrieves total number of messages ∀ users. 
+   * @return messages the total number of messages ∀ users
+   */
+  public int getTotalMessageCount(){
+    Query query = new Query("Message");
+    PreparedQuery results = datastore.prepare(query);
+    return results.countEntities(FetchOptions.Builder.withLimit(1000));
+    /**
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/data/Datastore.java]: nit: remove this comment

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 4]
```diff
@@ -27,57 +28,69 @@
 import java.util.List;
 import java.util.UUID;
 
+
 /** Provides access to the data stored in Datastore. */
-public class Datastore {
+public class Datastore{
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/data/Datastore.java]: I'm going to add a style checker for the google java style guide which will enforce having the space before an opening brace so lets leave these in the file.

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 5]
```diff
@@ -27,57 +28,69 @@
...
    */
-  public List<Message> getMessages(String user) {
+  public List<Message> getMessages(String user){
     List<Message> messages = new ArrayList<>();
 
-    Query query =
-        new Query("Message")
-            .setFilter(new Query.FilterPredicate("user", FilterOperator.EQUAL, user))
-            .addSort("timestamp", SortDirection.DESCENDING);
+    Query query = new Query("Message").setFilter(new Query.FilterPredicate("user", FilterOperator.EQUAL, user)).addSort("timestamp", SortDirection.DESCENDING);
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/data/Datastore.java]: Lets leave the formatting of this line. It's now way longer than 120 characters which breaks a lot of tools (including github in this case).

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 6]
```diff
@@ -16,6 +16,7 @@
 
 package com.google.codeu.data;
 
+import com.google.appengine.api.datastore.FetchOptions;   //needed for function w/ total message count
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/data/Datastore.java]: nit: put imports in alphabetical order

Comment (dgalbraith33) [inline src/main/java/com/google/codeu/data/Datastore.java]: nit: let's remove this comment. We generally only comment imports if its very difficult to figure out what they're for. In this case the reader can simply search for "FetchOptions" and see where it's used.

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
+/**
+ * SERVELET FOR STATS PAGE
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: nit: spelling SERVLET

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+package com.google.codeu.servlets;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+
+//Modified to output message data as JSON
+import com.google.codeu.data.Datastore;
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: nit: drop comment and put these with the other imports (in alphabetical order)

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import com.google.gson.JsonObject;
+
+ //Handles fetching site statistics
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: nit: drop leading space before comment

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+@WebServlet("/stats")
+public class StatsPageServlet extends HttpServlet{
+
+  private Datastore datastore;
+
+  //Automatically called when servlet is first created
+  @Override
+  public void init(){
+    datastore = new Datastore();  //Creates + stores DS instance
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: nit: drop the comment on this line (it repeats what the code does)

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    int message_count = 0;
+    
+    response.setContentType("application/json");
+    message_count = datastore.getTotalMessageCount();
+    JsonObject jsonObject = new JsonObject();
+    jsonObject.addProperty("messageCount", message_count);
+
+    response.getOutputStream().println(jsonObject.toString());
+ }
+}
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: nit: add a newline at the end of the file

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+  //Automatically called when servlet is first created
+  @Override
+  public void init(){
+    datastore = new Datastore();  //Creates + stores DS instance
+  }
+
+  // Responds with site stats in JSON
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException{
+    int message_count = 0;
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: nit: no need to define this at the top of the function. You can simply do "int message_count = datastore.getTotalMessageCount();" on that line

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 13]
```diff
@@ -27,57 +28,69 @@
...
+      try{
         String idString = entity.getKey().getName();
         UUID id = UUID.fromString(idString);
-        String text = (String) entity.getProperty("text");
-        long timestamp = (long) entity.getProperty("timestamp");
-        String recipient = (String) entity.getProperty("recipient");
-        Message message = new Message(id, user, text, timestamp, recipient);
-        messages.add(message);
-      } catch (Exception e) {
+        String message_text = (String)entity.getProperty("text");
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/data/Datastore.java]: I'm alright with changing variable names in general to make them more descriptive. However it's better to do that when 
a) you are already changing that code in a part of a PR
b) you are making a specific PR for the purpose of renaming them.

I'm ok with leaving these changes in this one - just a not for the future.

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 14]
```diff
@@ -27,57 +28,69 @@
...
-        new Query("Message")
-            .setFilter(new Query.FilterPredicate("user", FilterOperator.EQUAL, user))
-            .addSort("timestamp", SortDirection.DESCENDING);
+    Query query = new Query("Message").setFilter(new Query.FilterPredicate("user", FilterOperator.EQUAL, user)).addSort("timestamp", SortDirection.DESCENDING);
+    
     PreparedQuery results = datastore.prepare(query);
-
-    for (Entity entity : results.asIterable()) {
-      try {
+    for(Entity entity : results.asIterable()){
```
Comment (dgalbraith33) [inline src/main/java/com/google/codeu/data/Datastore.java]: It looks like a lot of the space separators were dropped in this block. Let's leave those for style purposes (and to stay consistent with the rest of the files.

**Comment by GigaMatt** [comment]
I hope I didn't do anything too bad by adding the copyright info to the StatsPageServlet.java file.

I noticed all of the other Servlet files had them since we're contributing them, I don't want to fall on the wrong(er) side of copyrights, so please let me know if I should remove it/fix the formatting.

File: src/main/java/com/google/codeu/servlets/StatsPageServlet.java (hunk 9 above)
Comment (dgalbraith33) [reply src/main/java/com/google/codeu/servlets/StatsPageServlet.java]: Sorry I was unclear here. You can leave the newline above it. Theres a space on the line before the comment. 

" //Handles"

**Pull request merged**